### ISSUE TITLE
Does loss dict keep the same order in different processes?

### ISSUE THREAD

**Issue by yelantingfeng** [facebookresearch/maskrcnn-benchmark#309]
## ❓ Questions and Help
In maskrcnn_benchmark.engine.trainer, the function `reduce_loss_dict` reduce all losses in loss_dict to rank 0. 
`
def reduce_loss_dict(loss_dict):
    """
    Reduce the loss dictionary from all processes so that process with rank
    0 has the averaged results. Returns a dict with the same fields as
    loss_dict, after reduction.
    """
    world_size = get_world_size()
    if world_size < 2:
        return loss_dict
    with torch.no_grad():
        loss_names = []
        all_losses = []
        for k, v in loss_dict.items():
            loss_names.append(k)
            all_losses.append(v)
        all_losses = torch.stack(all_losses, dim=0)
        dist.reduce(all_losses, dst=0)
        if dist.get_rank() == 0:
            # only main process gets accumulated, so only divide by
            # world_size in this case
            all_losses /= world_size
        reduced_losses = {k: v for k, v in zip(loss_names, all_losses)}
    return reduced_losses
`

It uses loss_dict.items() and I think the order of plain dictionary object keys is not deterministic. Would it be better if we use OrderDict?

**Comment by fmassa** [comment]
Yes, ordered dict would be safer.
Another possibility would be to sort loss_names and use this order to sort all_losses.
Can you send a PR with the change?

**Comment by yelantingfeng** [comment]
Yes, I can do that.
